libraries/AP_Scripting/applets/RockBlock.lua

```diff
@@ -0,0 +1,888 @@
+        -- The checksum is the least significant 2-bytes of the summation of the entire SBD message
+        local SUM = 0
+
+        for idx = 1, #bytes do SUM = SUM + bytes:byte(idx) end
+
+        local SUM_H = (SUM & 0xFF << 8) >> 8
+        local SUM_L = SUM & 0xFF
+
+        if self.debug_text then
+            gcs:send_text(3, "Rockblock: Modem CRC: " .. string.char(SUM_H) .. " " ..
+                              string.char(SUM_L))
+        end
+
+        return string.char(SUM_H), string.char(SUM_L)
+    end
+
+    function self.nicestring(instr)
+        -- make any strings printable to GCS (constrain to ASCII range)
+        local retstr = ""
+        for i = 1, #instr do
+            local c = string.byte(instr:sub(i, i))
+            if c < 0x20 or c > 0x7E then c = 0x5F end
+            retstr = retstr .. string.char(c)
+        end
+
+        return tostring(retstr)
+    end
+    
+    function self.checkmodem()
+        --- send detect command to modem every 10 sec if not detected
+        if (millis():tofloat() * 0.001) - self.last_modem_status_check > 10 then
+            table.insert(_modem_to_send, _AT_query)
+            self.last_modem_status_check = millis():tofloat() * 0.001
+        end
+    end
+    
+    function self.getnextcommandtosend()
+        --- get the next string to send to the modem
+        if #_modem_to_send == 0 then
+            return nil
+        end
+        
+        local ret = _modem_to_send[1]
+        if self.debug_text then
+            gcs:send_text(3, "Rockblock: Sent to modem: " .. self.nicestring(_modem_to_send[1]))
+        end
+        table.remove(_modem_to_send, 1)
+        
+        return ret
+    end
+    
+    function self.loadtxbuffer(pkt)
+        -- Load the transmit buffer of the modem
+        -- Send as binary data, plus checksum
+        local highByte, lowByte = self.checksum(pkt)
+        table.insert(_modem_to_send, _AT_load_tx_buffer .. tostring(#pkt) .. '\r')
+        table.insert(_modem_to_send, pkt)
+        table.insert(_modem_to_send, highByte .. lowByte .. "\r")
+
+        self.time_last_tx = millis():tofloat() * 0.001
+        table.insert(_modem_to_send, _AT_mailbox_check)
+        self.is_transmitting = true
+    end
+    
+    -- return the instance
+    return self
+end
+
+-- Transmitted HIGH_LATENCY2 packet
+local hl2 = {}
+hl2.timestamp = 0
+hl2.latitude = 0
+hl2.longitude = 0
+hl2.custom_mode = 0
+hl2.altitude = 0
+hl2.target_altitude = 0
+hl2.target_distance = 0
+hl2.wp_num = 0
+hl2.failure_flags = 0
+hl2.type = gcs:frame_type()
+hl2.autopilot = 3 -- MAV_AUTOPILOT_ARDUPILOTMEGA
+hl2.heading = 0
+hl2.target_heading = 0
+hl2.throttle = 0
+hl2.airspeed = 0
+hl2.airspeed_sp = 0
+hl2.groundspeed = 0
+hl2.windspeed = 0
+hl2.wind_heading = 0
+hl2.eph = 0
+hl2.epv = 0
+hl2.temperature_air = 0
+hl2.climb_rate = 0
+hl2.battery = 0
+hl2.custom0 = 1 -- MAV_MODE_FLAG_CUSTOM_MODE_ENABLED
+hl2.custom1 = 0
+hl2.custom2 = 0
+
+function wrap_360(angle)
```

Comment:
Instead of reinventing the wheel, why not just expose all the C++ functions to scripting?